### PR TITLE
[gluten-core] Support repeat function

### PR DESCRIPTION
**Pull request by KevinyhZou**
## What changes were proposed in this pull request?
support string repeat function in gluten, the releated issue: https://github.com/oap-project/gluten/issues/974

the clickhouse backend support repeat function pr : https://github.com/Kyligence/ClickHouse/pull/298 
https://github.com/ClickHouse/ClickHouse/pull/47134

## How was this patch tested?
test by manual

### CODE REVIEW

**Comment by jinchengchenghh** [comment]
Can you fix the error in CI? You can search `*** F` in downloaded log.
One of the error is
```
2023-02-21T08:34:36.1840402Z 23/02/21 16:34:36 ERROR GlutenSQLQueryTestSuite: Error using configs: 
2023-02-21T08:34:36.2222173Z [31m- charvarchar.sql *** FAILED ***[0m
2023-02-21T08:34:36.2223347Z [31m  charvarchar.sql[0m
2023-02-21T08:34:36.2225115Z [31m  Expected "struct<[repeat(c7, 2):string,repeat(c8, 2):string,repeat(v, 2):string,repeat(s, 2):string]>", but got "struct<[]>" Schema did not match for query #69[0m
2023-02-21T08:34:36.2227033Z [31m  select repeat(c7, 2), repeat(c8, 2), repeat(v, 2), repeat(s, 2) from char_tbl4: -- !query[0m
2023-02-21T08:34:36.2228589Z [31m  select repeat(c7, 2), repeat(c8, 2), repeat(v, 2), repeat(s, 2) from char_tbl4[0m
2023-02-21T08:34:36.2229719Z [31m  -- !query schema[0m
2023-02-21T08:34:36.2230512Z [31m  struct<>[0m
2023-02-21T08:34:36.2231419Z [31m  -- !query output[0m
2023-02-21T08:34:36.2232489Z [31m  java.lang.RuntimeException[0m
2023-02-21T08:34:36.2233486Z [31m  Exception: VeloxUserError[0m
2023-02-21T08:34:36.2234498Z [31m  Error Source: USER[0m
2023-02-21T08:34:36.2235575Z [31m  Reason: scalar type has no children[0m
2023-02-21T08:34:36.2236496Z [31m  Retriable: False[0m
2023-02-21T08:34:36.2237543Z [31m  Context: repeat(n0_0, 2:INTEGER)[0m
2023-02-21T08:34:36.2238680Z [31m  Top-Level Context: Same as context.[0m
2023-02-21T08:34:36.2239654Z [31m  Function: [0m
2023-02-21T08:34:36.2240447Z [31m  File: UNKNOWN[0m
2023-02-21T08:34:36.2241356Z [31m  Stack trace:[0m
2023-02-21T08:34:36.2243024Z [31m  Stack trace has been disabled.Use --velox_exception_user_stacktrace=true to enable it. (GlutenSQLQueryTestSuite.scala:723)[0m
2023-02-21T08:34:36.2967492Z [32m- change-column.sql[0m
2023-02-21T08:34:36.9106257Z [32m- try_arithmetic.sql[0m
```

**Comment by KevinyhZou** [comment]
> Can you fix the error in CI? You can search `*** F` in downloaded log. One of the error is
> 
> ```
> 2023-02-21T08:34:36.1840402Z 23/02/21 16:34:36 ERROR GlutenSQLQueryTestSuite: Error using configs: 
> 2023-02-21T08:34:36.2222173Z �[31m- charvarchar.sql *** FAILED ***�[0m
> 2023-02-21T08:34:36.2223347Z �[31m  charvarchar.sql�[0m
> 2023-02-21T08:34:36.2225115Z �[31m  Expected "struct<[repeat(c7, 2):string,repeat(c8, 2):string,repeat(v, 2):string,repeat(s, 2):string]>", but got "struct<[]>" Schema did not match for query #69�[0m
> 2023-02-21T08:34:36.2227033Z �[31m  select repeat(c7, 2), repeat(c8, 2), repeat(v, 2), repeat(s, 2) from char_tbl4: -- !query�[0m
> 2023-02-21T08:34:36.2228589Z �[31m  select repeat(c7, 2), repeat(c8, 2), repeat(v, 2), repeat(s, 2) from char_tbl4�[0m
> 2023-02-21T08:34:36.2229719Z �[31m  -- !query schema�[0m
> 2023-02-21T08:34:36.2230512Z �[31m  struct<>�[0m
> 2023-02-21T08:34:36.2231419Z �[31m  -- !query output�[0m
> 2023-02-21T08:34:36.2232489Z �[31m  java.lang.RuntimeException�[0m
> 2023-02-21T08:34:36.2233486Z �[31m  Exception: VeloxUserError�[0m
> 2023-02-21T08:34:36.2234498Z �[31m  Error Source: USER�[0m
> 2023-02-21T08:34:36.2235575Z �[31m  Reason: scalar type has no children�[0m
> 2023-02-21T08:34:36.2236496Z �[31m  Retriable: False�[0m
> 2023-02-21T08:34:36.2237543Z �[31m  Context: repeat(n0_0, 2:INTEGER)�[0m
> 2023-02-21T08:34:36.2238680Z �[31m  Top-Level Context: Same as context.�[0m
> 2023-02-21T08:34:36.2239654Z �[31m  Function: �[0m
> 2023-02-21T08:34:36.2240447Z �[31m  File: UNKNOWN�[0m
> 2023-02-21T08:34:36.2241356Z �[31m  Stack trace:�[0m
> 2023-02-21T08:34:36.2243024Z �[31m  Stack trace has been disabled.Use --velox_exception_user_stacktrace=true to enable it. (GlutenSQLQueryTestSuite.scala:723)�[0m
> 2023-02-21T08:34:36.2967492Z �[32m- change-column.sql�[0m
> 2023-02-21T08:34:36.9106257Z �[32m- try_arithmetic.sql�[0m
> ```

OK

File: gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala [hunk 1]
```diff
@@ -36,6 +36,13 @@ class BinaryExpressionTransformer(
     original: Expression)
   extends ExpressionTransformer with Logging {
   override def doTransform(args: java.lang.Object): ExpressionNode = {
+    if (ExpressionMappings.REPEAT.equalsIgnoreCase(original.prettyName) &&
+      !BackendsApiManager.getBackendName.equalsIgnoreCase(GlutenConfig.GLUTEN_CLICKHOUSE_BACKEND)) {
+      val errMsg = String.format("%s or %s is not currently supported for non clickhouse backend.",
+        original.getClass, original)
+      logWarning(errMsg)
+      throw new UnsupportedOperationException(errMsg)
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala:44]: How about adding this function into below blacklist for Velox backend to save the extra check here?
[VeloxExpressionUtil.scala#L31](https://github.com/oap-project/gluten/blob/main/backends-velox/src/main/scala/io/glutenproject/utils/VeloxExpressionUtil.scala#L31)

Comment (KevinyhZou) [reply gluten-core/src/main/scala/io/glutenproject/expression/BinaryExpressionTransformer.scala:44]: OK

**Review by rui-mo — APPROVED** [review]

**Pull request merged**